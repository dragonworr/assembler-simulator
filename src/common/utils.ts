import { produce } from 'immer'

export type NullablePartial<T> = {
  [P in keyof T]?: T[P] | undefined
}

export type Head<T extends unknown[]> = T extends [...infer Head, unknown] ? Head : unknown[]

export const sign8 = (unsigned: number): number =>
  unsigned >= 0x80 ? unsigned - 0xff - 1 : unsigned

export const unsign8 = (signed: number): number => (signed < 0 ? signed + 0xff + 1 : signed)

export const hexToDec = (str: string): number => Number.parseInt(str, 16)

export const decToHex = (num: number): string => num.toString(16).padStart(2, '0').toUpperCase()

export const stringToASCII = (str: string): number[] =>
  str.split('').map(char => char.charCodeAt(0))

export const trimBrackets = (str: string): string => str.replace(/^\[(.*)]$/, '$1')

export const splitArrayPerChunk = <T>(arr: T[], perChunk: number): T[][] =>
  arr.reduce<T[][]>((res, val, index) => {
    const chunkIndex = Math.floor(index / perChunk)
    return produce(res, (draft: T[][]) => {
      draft[chunkIndex] = [...(draft[chunkIndex] ?? []), val]
    })
  }, [])

export const call = <T>(fn: () => T): T => fn()
