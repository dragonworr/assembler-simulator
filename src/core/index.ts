export { assemble } from './assembler'
export * from './memory'
export * from './cpu'
