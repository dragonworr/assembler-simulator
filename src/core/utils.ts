export const hexToDec = (str: string): number => Number.parseInt(str, 16)

export const decToHex = (num: number): string => num.toString(16).padStart(2, '0').toUpperCase()

export const stringToASCII = (str: string): number[] =>
  str.split('').map(char => char.charCodeAt(0))

export const trimBrackets = (str: string): string => str.replace(/^\[(.*)]$/, '$1')

export const splitUint8ArrayPerChunk = (arr: Uint8Array, perChunk: number): number[][] =>
  arr.reduce<number[][]>((matrix, value, index) => {
    const chunkIndex = Math.floor(index / perChunk)
    matrix[chunkIndex] = matrix[chunkIndex] ?? []
    matrix[chunkIndex].push(value)
    return matrix
  }, [])
