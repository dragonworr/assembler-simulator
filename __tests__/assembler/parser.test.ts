import type { Statement } from '../../src/core/assembler'
import { tokenize, parse as _parse } from '../../src/core/assembler'
import { expectError } from '../utils'

const parse = (input: string): Statement[] => _parse(tokenize(input))

describe('parser', () => {
  it('should parse instruction with no operand', () => {
    expect(
      parse(`
pushf
popf
ret
iret
halt
nop
clo
cli
sti
end
`)
    ).toMatchSnapshot()
  })

  it('should parse instruction with 1 operand', () => {
    expect(
      parse(`
start:
inc al
dec bl
not cl
rol dl
ror al
shl bl
shr cl
jmp start
jz start
jnz start
js start
jns start
jo start
jno start
push al
pop bl
call 0
int 1
in 2
out 3
org 4
db 5
db "67"
end
`)
    ).toMatchSnapshot()
  })

  it('should parse direct arithmetic instruction', () => {
    expect(
      parse(`
add al, bl
sub cl, dl
mul al, bl
div cl, dl
mod al, bl
and cl, dl
or al, bl
xor cl, dl
end
`)
    ).toMatchSnapshot()
  })

  it('should parse immediate arithmetic instruction', () => {
    expect(
      parse(`
add al, 01
sub bl, 02
mul cl, 03
div dl, 04
mod al, 01
and bl, 02
or cl, 03
xor dl, 04
end
`)
    ).toMatchSnapshot()
  })

  it('should parse move instruction', () => {
    expect(
      parse(`
mov al, 01
mov bl, [02]
mov [03], cl
mov dl, [al]
mov [bl], cl
end
`)
    ).toMatchSnapshot()
  })

  it('should parse compare instruction', () => {
    expect(
      parse(`
cmp al, bl
cmp cl, 01
cmp dl, [02]
end
`)
    ).toMatchSnapshot()
  })

  it('should throw InvalidLabelError', () => {
    expectError(() => {
      parse('!done: end')
    }, 'Label should start with a charactor or underscore: !done')

    expectError(() => {
      parse(': end')
    }, 'Label should start with a charactor or underscore: ')

    expectError(() => {
      parse('jmp !start end')
    }, 'Label should start with a charactor or underscore: !start')
  })

  const tokensKnown = [',', '01', 'al', '[bl]', '"cl"']

  it('should throw StatementError without label', () => {
    tokensKnown.forEach(token => {
      expectError(() => {
        parse(`${token} end`)
      }, `Expected label or instruction: ${token}`)
    })
  })

  it('should throw StatementError with label', () => {
    tokensKnown.forEach(token => {
      expectError(() => {
        parse(`start: ${token} end`)
      }, `Expected instruction: ${token}`)
    })
  })

  const MISSING_END_ERROR_MSG = 'Expected END at the end of the source code'

  it('should throw MissingEndError when missing operand', () => {
    expectError(() => {
      parse('mov')
    }, MISSING_END_ERROR_MSG)
  })

  it('should throw MissingEndError when missing comma', () => {
    expectError(() => {
      parse('mov al')
    }, MISSING_END_ERROR_MSG)
  })

  it('should throw MissingEndError when missing instruction', () => {
    expectError(() => {
      parse('start:')
    }, MISSING_END_ERROR_MSG)
  })

  it('should throw MissingEndError when parsing is completed', () => {
    expectError(() => {
      parse('mov al, 01')
    }, MISSING_END_ERROR_MSG)
  })

  it('should throw AddressError when address is invalid', () => {
    expectError(() => {
      parse('mov [gg], al')
    }, 'Expected a number or register: gg')
  })

  it('should throw AddressError when address is empty', () => {
    expectError(() => {
      parse('mov [], al')
    }, 'Expected a number or register: ]')
  })

  it('should throw InvalidNumberError', () => {
    expectError(() => {
      parse('mov al, 100')
    }, 'Number should be hexadecimal and less than or equal to FF: 100')
  })

  it('should throw InvalidNumberError when parsing address', () => {
    expectError(() => {
      parse('mov [100], al')
    }, 'Number should be hexadecimal and less than or equal to FF: 100')
  })

  it('should throw InvalidNumberError when parsing non-digit number', () => {
    expectError(() => {
      parse('mov al, fff')
    }, 'Number should be hexadecimal and less than or equal to FF: fff')
  })

  it('should throw OperandTypeError with one expected type', () => {
    expectError(() => {
      parse('inc 01')
    }, 'Expected register: 01')
  })

  it('should throw OperandTypeError with more than one expected types', () => {
    expectError(() => {
      parse('mov 01')
    }, 'Expected register, address or register address: 01')
  })

  describe('move instruction', () => {
    it('should throw OperandTypeError when parsing second operand', () => {
      expectError(() => {
        parse('mov al, "bl"')
      }, 'Expected number, address or register address: "bl"')

      expectError(() => {
        parse('mov [al], "01"')
      }, 'Expected register: "01"')

      expectError(() => {
        parse('mov [01], "02"')
      }, 'Expected register: "02"')
    })

    it('should throw OperandTypeError when the types of the two operands are not allowed', () => {
      expectError(() => {
        parse('mov al, bl')
      }, 'Expected number, address or register address: bl')

      expectError(() => {
        parse('mov [01], 02')
      }, 'Expected register: 02')

      expectError(() => {
        parse('mov [al], 01')
      }, 'Expected register: 01')
    })
  })

  it('should throw MissingCommaError', () => {
    expectError(() => {
      parse('mov al 01')
    }, 'Expected comma: 01')
  })
})
