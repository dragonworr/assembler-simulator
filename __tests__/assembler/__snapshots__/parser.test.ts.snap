// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser should parse compare instruction 1`] = `
Array [
  Object {
    "instruction": Object {
      "mnemonic": "CMP",
      "opcode": 218,
      "range": Array [1, 4],
    },
    "label": null,
    "machineCodes": Array [218, 0, 1],
    "operands": Array [
      Object {
        "range": Array [5, 7],
        "raw": "al",
        "rawValue": "AL",
        "type": "Register",
        "value": 0,
      },
      Object {
        "range": Array [9, 11],
        "raw": "bl",
        "rawValue": "BL",
        "type": "Register",
        "value": 1,
      },
    ],
    "range": Array [1, 11],
  },
  Object {
    "instruction": Object {
      "mnemonic": "CMP",
      "opcode": 219,
      "range": Array [12, 15],
    },
    "label": null,
    "machineCodes": Array [219, 2, 1],
    "operands": Array [
      Object {
        "range": Array [16, 18],
        "raw": "cl",
        "rawValue": "CL",
        "type": "Register",
        "value": 2,
      },
      Object {
        "range": Array [20, 22],
        "raw": "01",
        "rawValue": "01",
        "type": "Number",
        "value": 1,
      },
    ],
    "range": Array [12, 22],
  },
  Object {
    "instruction": Object {
      "mnemonic": "CMP",
      "opcode": 220,
      "range": Array [23, 26],
    },
    "label": null,
    "machineCodes": Array [220, 3, 2],
    "operands": Array [
      Object {
        "range": Array [27, 29],
        "raw": "dl",
        "rawValue": "DL",
        "type": "Register",
        "value": 3,
      },
      Object {
        "range": Array [31, 35],
        "raw": "[02]",
        "rawValue": "02",
        "type": "Address",
        "value": 2,
      },
    ],
    "range": Array [23, 35],
  },
  Object {
    "instruction": Object {
      "mnemonic": "END",
      "opcode": 0,
      "range": Array [36, 39],
    },
    "label": null,
    "machineCodes": Array [0],
    "operands": Array [],
    "range": Array [36, 39],
  },
]
`;

exports[`parser should parse direct arithmetic instruction 1`] = `
Array [
  Object {
    "instruction": Object {
      "mnemonic": "ADD",
      "opcode": 160,
      "range": Array [1, 4],
    },
    "label": null,
    "machineCodes": Array [160, 0, 1],
    "operands": Array [
      Object {
        "range": Array [5, 7],
        "raw": "al",
        "rawValue": "AL",
        "type": "Register",
        "value": 0,
      },
      Object {
        "range": Array [9, 11],
        "raw": "bl",
        "rawValue": "BL",
        "type": "Register",
        "value": 1,
      },
    ],
    "range": Array [1, 11],
  },
  Object {
    "instruction": Object {
      "mnemonic": "SUB",
      "opcode": 161,
      "range": Array [12, 15],
    },
    "label": null,
    "machineCodes": Array [161, 2, 3],
    "operands": Array [
      Object {
        "range": Array [16, 18],
        "raw": "cl",
        "rawValue": "CL",
        "type": "Register",
        "value": 2,
      },
      Object {
        "range": Array [20, 22],
        "raw": "dl",
        "rawValue": "DL",
        "type": "Register",
        "value": 3,
      },
    ],
    "range": Array [12, 22],
  },
  Object {
    "instruction": Object {
      "mnemonic": "MUL",
      "opcode": 162,
      "range": Array [23, 26],
    },
    "label": null,
    "machineCodes": Array [162, 0, 1],
    "operands": Array [
      Object {
        "range": Array [27, 29],
        "raw": "al",
        "rawValue": "AL",
        "type": "Register",
        "value": 0,
      },
      Object {
        "range": Array [31, 33],
        "raw": "bl",
        "rawValue": "BL",
        "type": "Register",
        "value": 1,
      },
    ],
    "range": Array [23, 33],
  },
  Object {
    "instruction": Object {
      "mnemonic": "DIV",
      "opcode": 163,
      "range": Array [34, 37],
    },
    "label": null,
    "machineCodes": Array [163, 2, 3],
    "operands": Array [
      Object {
        "range": Array [38, 40],
        "raw": "cl",
        "rawValue": "CL",
        "type": "Register",
        "value": 2,
      },
      Object {
        "range": Array [42, 44],
        "raw": "dl",
        "rawValue": "DL",
        "type": "Register",
        "value": 3,
      },
    ],
    "range": Array [34, 44],
  },
  Object {
    "instruction": Object {
      "mnemonic": "MOD",
      "opcode": 166,
      "range": Array [45, 48],
    },
    "label": null,
    "machineCodes": Array [166, 0, 1],
    "operands": Array [
      Object {
        "range": Array [49, 51],
        "raw": "al",
        "rawValue": "AL",
        "type": "Register",
        "value": 0,
      },
      Object {
        "range": Array [53, 55],
        "raw": "bl",
        "rawValue": "BL",
        "type": "Register",
        "value": 1,
      },
    ],
    "range": Array [45, 55],
  },
  Object {
    "instruction": Object {
      "mnemonic": "AND",
      "opcode": 170,
      "range": Array [56, 59],
    },
    "label": null,
    "machineCodes": Array [170, 2, 3],
    "operands": Array [
      Object {
        "range": Array [60, 62],
        "raw": "cl",
        "rawValue": "CL",
        "type": "Register",
        "value": 2,
      },
      Object {
        "range": Array [64, 66],
        "raw": "dl",
        "rawValue": "DL",
        "type": "Register",
        "value": 3,
      },
    ],
    "range": Array [56, 66],
  },
  Object {
    "instruction": Object {
      "mnemonic": "OR",
      "opcode": 171,
      "range": Array [67, 69],
    },
    "label": null,
    "machineCodes": Array [171, 0, 1],
    "operands": Array [
      Object {
        "range": Array [70, 72],
        "raw": "al",
        "rawValue": "AL",
        "type": "Register",
        "value": 0,
      },
      Object {
        "range": Array [74, 76],
        "raw": "bl",
        "rawValue": "BL",
        "type": "Register",
        "value": 1,
      },
    ],
    "range": Array [67, 76],
  },
  Object {
    "instruction": Object {
      "mnemonic": "XOR",
      "opcode": 172,
      "range": Array [77, 80],
    },
    "label": null,
    "machineCodes": Array [172, 2, 3],
    "operands": Array [
      Object {
        "range": Array [81, 83],
        "raw": "cl",
        "rawValue": "CL",
        "type": "Register",
        "value": 2,
      },
      Object {
        "range": Array [85, 87],
        "raw": "dl",
        "rawValue": "DL",
        "type": "Register",
        "value": 3,
      },
    ],
    "range": Array [77, 87],
  },
  Object {
    "instruction": Object {
      "mnemonic": "END",
      "opcode": 0,
      "range": Array [88, 91],
    },
    "label": null,
    "machineCodes": Array [0],
    "operands": Array [],
    "range": Array [88, 91],
  },
]
`;

exports[`parser should parse immediate arithmetic instruction 1`] = `
Array [
  Object {
    "instruction": Object {
      "mnemonic": "ADD",
      "opcode": 176,
      "range": Array [1, 4],
    },
    "label": null,
    "machineCodes": Array [176, 0, 1],
    "operands": Array [
      Object {
        "range": Array [5, 7],
        "raw": "al",
        "rawValue": "AL",
        "type": "Register",
        "value": 0,
      },
      Object {
        "range": Array [9, 11],
        "raw": "01",
        "rawValue": "01",
        "type": "Number",
        "value": 1,
      },
    ],
    "range": Array [1, 11],
  },
  Object {
    "instruction": Object {
      "mnemonic": "SUB",
      "opcode": 177,
      "range": Array [12, 15],
    },
    "label": null,
    "machineCodes": Array [177, 1, 2],
    "operands": Array [
      Object {
        "range": Array [16, 18],
        "raw": "bl",
        "rawValue": "BL",
        "type": "Register",
        "value": 1,
      },
      Object {
        "range": Array [20, 22],
        "raw": "02",
        "rawValue": "02",
        "type": "Number",
        "value": 2,
      },
    ],
    "range": Array [12, 22],
  },
  Object {
    "instruction": Object {
      "mnemonic": "MUL",
      "opcode": 178,
      "range": Array [23, 26],
    },
    "label": null,
    "machineCodes": Array [178, 2, 3],
    "operands": Array [
      Object {
        "range": Array [27, 29],
        "raw": "cl",
        "rawValue": "CL",
        "type": "Register",
        "value": 2,
      },
      Object {
        "range": Array [31, 33],
        "raw": "03",
        "rawValue": "03",
        "type": "Number",
        "value": 3,
      },
    ],
    "range": Array [23, 33],
  },
  Object {
    "instruction": Object {
      "mnemonic": "DIV",
      "opcode": 179,
      "range": Array [34, 37],
    },
    "label": null,
    "machineCodes": Array [179, 3, 4],
    "operands": Array [
      Object {
        "range": Array [38, 40],
        "raw": "dl",
        "rawValue": "DL",
        "type": "Register",
        "value": 3,
      },
      Object {
        "range": Array [42, 44],
        "raw": "04",
        "rawValue": "04",
        "type": "Number",
        "value": 4,
      },
    ],
    "range": Array [34, 44],
  },
  Object {
    "instruction": Object {
      "mnemonic": "MOD",
      "opcode": 182,
      "range": Array [45, 48],
    },
    "label": null,
    "machineCodes": Array [182, 0, 1],
    "operands": Array [
      Object {
        "range": Array [49, 51],
        "raw": "al",
        "rawValue": "AL",
        "type": "Register",
        "value": 0,
      },
      Object {
        "range": Array [53, 55],
        "raw": "01",
        "rawValue": "01",
        "type": "Number",
        "value": 1,
      },
    ],
    "range": Array [45, 55],
  },
  Object {
    "instruction": Object {
      "mnemonic": "AND",
      "opcode": 186,
      "range": Array [56, 59],
    },
    "label": null,
    "machineCodes": Array [186, 1, 2],
    "operands": Array [
      Object {
        "range": Array [60, 62],
        "raw": "bl",
        "rawValue": "BL",
        "type": "Register",
        "value": 1,
      },
      Object {
        "range": Array [64, 66],
        "raw": "02",
        "rawValue": "02",
        "type": "Number",
        "value": 2,
      },
    ],
    "range": Array [56, 66],
  },
  Object {
    "instruction": Object {
      "mnemonic": "OR",
      "opcode": 187,
      "range": Array [67, 69],
    },
    "label": null,
    "machineCodes": Array [187, 2, 3],
    "operands": Array [
      Object {
        "range": Array [70, 72],
        "raw": "cl",
        "rawValue": "CL",
        "type": "Register",
        "value": 2,
      },
      Object {
        "range": Array [74, 76],
        "raw": "03",
        "rawValue": "03",
        "type": "Number",
        "value": 3,
      },
    ],
    "range": Array [67, 76],
  },
  Object {
    "instruction": Object {
      "mnemonic": "XOR",
      "opcode": 188,
      "range": Array [77, 80],
    },
    "label": null,
    "machineCodes": Array [188, 3, 4],
    "operands": Array [
      Object {
        "range": Array [81, 83],
        "raw": "dl",
        "rawValue": "DL",
        "type": "Register",
        "value": 3,
      },
      Object {
        "range": Array [85, 87],
        "raw": "04",
        "rawValue": "04",
        "type": "Number",
        "value": 4,
      },
    ],
    "range": Array [77, 87],
  },
  Object {
    "instruction": Object {
      "mnemonic": "END",
      "opcode": 0,
      "range": Array [88, 91],
    },
    "label": null,
    "machineCodes": Array [0],
    "operands": Array [],
    "range": Array [88, 91],
  },
]
`;

exports[`parser should parse instruction with 1 operand 1`] = `
Array [
  Object {
    "instruction": Object {
      "mnemonic": "INC",
      "opcode": 164,
      "range": Array [8, 11],
    },
    "label": Object {
      "identifier": "START",
      "range": Array [1, 6],
    },
    "machineCodes": Array [164, 0],
    "operands": Array [
      Object {
        "range": Array [12, 14],
        "raw": "al",
        "rawValue": "AL",
        "type": "Register",
        "value": 0,
      },
    ],
    "range": Array [8, 14],
  },
  Object {
    "instruction": Object {
      "mnemonic": "DEC",
      "opcode": 165,
      "range": Array [15, 18],
    },
    "label": null,
    "machineCodes": Array [165, 1],
    "operands": Array [
      Object {
        "range": Array [19, 21],
        "raw": "bl",
        "rawValue": "BL",
        "type": "Register",
        "value": 1,
      },
    ],
    "range": Array [15, 21],
  },
  Object {
    "instruction": Object {
      "mnemonic": "NOT",
      "opcode": 173,
      "range": Array [22, 25],
    },
    "label": null,
    "machineCodes": Array [173, 2],
    "operands": Array [
      Object {
        "range": Array [26, 28],
        "raw": "cl",
        "rawValue": "CL",
        "type": "Register",
        "value": 2,
      },
    ],
    "range": Array [22, 28],
  },
  Object {
    "instruction": Object {
      "mnemonic": "ROL",
      "opcode": 154,
      "range": Array [29, 32],
    },
    "label": null,
    "machineCodes": Array [154, 3],
    "operands": Array [
      Object {
        "range": Array [33, 35],
        "raw": "dl",
        "rawValue": "DL",
        "type": "Register",
        "value": 3,
      },
    ],
    "range": Array [29, 35],
  },
  Object {
    "instruction": Object {
      "mnemonic": "ROR",
      "opcode": 155,
      "range": Array [36, 39],
    },
    "label": null,
    "machineCodes": Array [155, 0],
    "operands": Array [
      Object {
        "range": Array [40, 42],
        "raw": "al",
        "rawValue": "AL",
        "type": "Register",
        "value": 0,
      },
    ],
    "range": Array [36, 42],
  },
  Object {
    "instruction": Object {
      "mnemonic": "SHL",
      "opcode": 156,
      "range": Array [43, 46],
    },
    "label": null,
    "machineCodes": Array [156, 1],
    "operands": Array [
      Object {
        "range": Array [47, 49],
        "raw": "bl",
        "rawValue": "BL",
        "type": "Register",
        "value": 1,
      },
    ],
    "range": Array [43, 49],
  },
  Object {
    "instruction": Object {
      "mnemonic": "SHR",
      "opcode": 157,
      "range": Array [50, 53],
    },
    "label": null,
    "machineCodes": Array [157, 2],
    "operands": Array [
      Object {
        "range": Array [54, 56],
        "raw": "cl",
        "rawValue": "CL",
        "type": "Register",
        "value": 2,
      },
    ],
    "range": Array [50, 56],
  },
  Object {
    "instruction": Object {
      "mnemonic": "JMP",
      "opcode": 192,
      "range": Array [57, 60],
    },
    "label": null,
    "machineCodes": Array [192],
    "operands": Array [
      Object {
        "range": Array [61, 66],
        "raw": "start",
        "rawValue": "START",
        "type": "Label",
        "value": undefined,
      },
    ],
    "range": Array [57, 66],
  },
  Object {
    "instruction": Object {
      "mnemonic": "JZ",
      "opcode": 193,
      "range": Array [67, 69],
    },
    "label": null,
    "machineCodes": Array [193],
    "operands": Array [
      Object {
        "range": Array [70, 75],
        "raw": "start",
        "rawValue": "START",
        "type": "Label",
        "value": undefined,
      },
    ],
    "range": Array [67, 75],
  },
  Object {
    "instruction": Object {
      "mnemonic": "JNZ",
      "opcode": 194,
      "range": Array [76, 79],
    },
    "label": null,
    "machineCodes": Array [194],
    "operands": Array [
      Object {
        "range": Array [80, 85],
        "raw": "start",
        "rawValue": "START",
        "type": "Label",
        "value": undefined,
      },
    ],
    "range": Array [76, 85],
  },
  Object {
    "instruction": Object {
      "mnemonic": "JS",
      "opcode": 195,
      "range": Array [86, 88],
    },
    "label": null,
    "machineCodes": Array [195],
    "operands": Array [
      Object {
        "range": Array [89, 94],
        "raw": "start",
        "rawValue": "START",
        "type": "Label",
        "value": undefined,
      },
    ],
    "range": Array [86, 94],
  },
  Object {
    "instruction": Object {
      "mnemonic": "JNS",
      "opcode": 196,
      "range": Array [95, 98],
    },
    "label": null,
    "machineCodes": Array [196],
    "operands": Array [
      Object {
        "range": Array [99, 104],
        "raw": "start",
        "rawValue": "START",
        "type": "Label",
        "value": undefined,
      },
    ],
    "range": Array [95, 104],
  },
  Object {
    "instruction": Object {
      "mnemonic": "JO",
      "opcode": 197,
      "range": Array [105, 107],
    },
    "label": null,
    "machineCodes": Array [197],
    "operands": Array [
      Object {
        "range": Array [108, 113],
        "raw": "start",
        "rawValue": "START",
        "type": "Label",
        "value": undefined,
      },
    ],
    "range": Array [105, 113],
  },
  Object {
    "instruction": Object {
      "mnemonic": "JNO",
      "opcode": 198,
      "range": Array [114, 117],
    },
    "label": null,
    "machineCodes": Array [198],
    "operands": Array [
      Object {
        "range": Array [118, 123],
        "raw": "start",
        "rawValue": "START",
        "type": "Label",
        "value": undefined,
      },
    ],
    "range": Array [114, 123],
  },
  Object {
    "instruction": Object {
      "mnemonic": "PUSH",
      "opcode": 224,
      "range": Array [124, 128],
    },
    "label": null,
    "machineCodes": Array [224, 0],
    "operands": Array [
      Object {
        "range": Array [129, 131],
        "raw": "al",
        "rawValue": "AL",
        "type": "Register",
        "value": 0,
      },
    ],
    "range": Array [124, 131],
  },
  Object {
    "instruction": Object {
      "mnemonic": "POP",
      "opcode": 225,
      "range": Array [132, 135],
    },
    "label": null,
    "machineCodes": Array [225, 1],
    "operands": Array [
      Object {
        "range": Array [136, 138],
        "raw": "bl",
        "rawValue": "BL",
        "type": "Register",
        "value": 1,
      },
    ],
    "range": Array [132, 138],
  },
  Object {
    "instruction": Object {
      "mnemonic": "CALL",
      "opcode": 202,
      "range": Array [139, 143],
    },
    "label": null,
    "machineCodes": Array [202, 0],
    "operands": Array [
      Object {
        "range": Array [144, 145],
        "raw": "0",
        "rawValue": "0",
        "type": "Number",
        "value": 0,
      },
    ],
    "range": Array [139, 145],
  },
  Object {
    "instruction": Object {
      "mnemonic": "INT",
      "opcode": 204,
      "range": Array [146, 149],
    },
    "label": null,
    "machineCodes": Array [204, 1],
    "operands": Array [
      Object {
        "range": Array [150, 151],
        "raw": "1",
        "rawValue": "1",
        "type": "Number",
        "value": 1,
      },
    ],
    "range": Array [146, 151],
  },
  Object {
    "instruction": Object {
      "mnemonic": "IN",
      "opcode": 240,
      "range": Array [152, 154],
    },
    "label": null,
    "machineCodes": Array [240, 2],
    "operands": Array [
      Object {
        "range": Array [155, 156],
        "raw": "2",
        "rawValue": "2",
        "type": "Number",
        "value": 2,
      },
    ],
    "range": Array [152, 156],
  },
  Object {
    "instruction": Object {
      "mnemonic": "OUT",
      "opcode": 241,
      "range": Array [157, 160],
    },
    "label": null,
    "machineCodes": Array [241, 3],
    "operands": Array [
      Object {
        "range": Array [161, 162],
        "raw": "3",
        "rawValue": "3",
        "type": "Number",
        "value": 3,
      },
    ],
    "range": Array [157, 162],
  },
  Object {
    "instruction": Object {
      "mnemonic": "ORG",
      "opcode": null,
      "range": Array [163, 166],
    },
    "label": null,
    "machineCodes": Array [4],
    "operands": Array [
      Object {
        "range": Array [167, 168],
        "raw": "4",
        "rawValue": "4",
        "type": "Number",
        "value": 4,
      },
    ],
    "range": Array [163, 168],
  },
  Object {
    "instruction": Object {
      "mnemonic": "DB",
      "opcode": null,
      "range": Array [169, 171],
    },
    "label": null,
    "machineCodes": Array [5],
    "operands": Array [
      Object {
        "range": Array [172, 173],
        "raw": "5",
        "rawValue": "5",
        "type": "Number",
        "value": 5,
      },
    ],
    "range": Array [169, 173],
  },
  Object {
    "instruction": Object {
      "mnemonic": "DB",
      "opcode": null,
      "range": Array [174, 176],
    },
    "label": null,
    "machineCodes": Array [54, 55],
    "operands": Array [
      Object {
        "range": Array [177, 181],
        "raw": "\\"67\\"",
        "rawValue": "67",
        "type": "String",
        "value": Array [54, 55],
      },
    ],
    "range": Array [174, 181],
  },
  Object {
    "instruction": Object {
      "mnemonic": "END",
      "opcode": 0,
      "range": Array [182, 185],
    },
    "label": null,
    "machineCodes": Array [0],
    "operands": Array [],
    "range": Array [182, 185],
  },
]
`;

exports[`parser should parse instruction with no operand 1`] = `
Array [
  Object {
    "instruction": Object {
      "mnemonic": "PUSHF",
      "opcode": 234,
      "range": Array [1, 6],
    },
    "label": null,
    "machineCodes": Array [234],
    "operands": Array [],
    "range": Array [1, 6],
  },
  Object {
    "instruction": Object {
      "mnemonic": "POPF",
      "opcode": 235,
      "range": Array [7, 11],
    },
    "label": null,
    "machineCodes": Array [235],
    "operands": Array [],
    "range": Array [7, 11],
  },
  Object {
    "instruction": Object {
      "mnemonic": "RET",
      "opcode": 203,
      "range": Array [12, 15],
    },
    "label": null,
    "machineCodes": Array [203],
    "operands": Array [],
    "range": Array [12, 15],
  },
  Object {
    "instruction": Object {
      "mnemonic": "IRET",
      "opcode": 205,
      "range": Array [16, 20],
    },
    "label": null,
    "machineCodes": Array [205],
    "operands": Array [],
    "range": Array [16, 20],
  },
  Object {
    "instruction": Object {
      "mnemonic": "HALT",
      "opcode": 0,
      "range": Array [21, 25],
    },
    "label": null,
    "machineCodes": Array [0],
    "operands": Array [],
    "range": Array [21, 25],
  },
  Object {
    "instruction": Object {
      "mnemonic": "NOP",
      "opcode": 255,
      "range": Array [26, 29],
    },
    "label": null,
    "machineCodes": Array [255],
    "operands": Array [],
    "range": Array [26, 29],
  },
  Object {
    "instruction": Object {
      "mnemonic": "CLO",
      "opcode": 254,
      "range": Array [30, 33],
    },
    "label": null,
    "machineCodes": Array [254],
    "operands": Array [],
    "range": Array [30, 33],
  },
  Object {
    "instruction": Object {
      "mnemonic": "CLI",
      "opcode": 253,
      "range": Array [34, 37],
    },
    "label": null,
    "machineCodes": Array [253],
    "operands": Array [],
    "range": Array [34, 37],
  },
  Object {
    "instruction": Object {
      "mnemonic": "STI",
      "opcode": 252,
      "range": Array [38, 41],
    },
    "label": null,
    "machineCodes": Array [252],
    "operands": Array [],
    "range": Array [38, 41],
  },
  Object {
    "instruction": Object {
      "mnemonic": "END",
      "opcode": 0,
      "range": Array [42, 45],
    },
    "label": null,
    "machineCodes": Array [0],
    "operands": Array [],
    "range": Array [42, 45],
  },
]
`;

exports[`parser should parse move instruction 1`] = `
Array [
  Object {
    "instruction": Object {
      "mnemonic": "MOV",
      "opcode": 208,
      "range": Array [1, 4],
    },
    "label": null,
    "machineCodes": Array [208, 0, 1],
    "operands": Array [
      Object {
        "range": Array [5, 7],
        "raw": "al",
        "rawValue": "AL",
        "type": "Register",
        "value": 0,
      },
      Object {
        "range": Array [9, 11],
        "raw": "01",
        "rawValue": "01",
        "type": "Number",
        "value": 1,
      },
    ],
    "range": Array [1, 11],
  },
  Object {
    "instruction": Object {
      "mnemonic": "MOV",
      "opcode": 209,
      "range": Array [12, 15],
    },
    "label": null,
    "machineCodes": Array [209, 1, 2],
    "operands": Array [
      Object {
        "range": Array [16, 18],
        "raw": "bl",
        "rawValue": "BL",
        "type": "Register",
        "value": 1,
      },
      Object {
        "range": Array [20, 24],
        "raw": "[02]",
        "rawValue": "02",
        "type": "Address",
        "value": 2,
      },
    ],
    "range": Array [12, 24],
  },
  Object {
    "instruction": Object {
      "mnemonic": "MOV",
      "opcode": 210,
      "range": Array [25, 28],
    },
    "label": null,
    "machineCodes": Array [210, 3, 2],
    "operands": Array [
      Object {
        "range": Array [29, 33],
        "raw": "[03]",
        "rawValue": "03",
        "type": "Address",
        "value": 3,
      },
      Object {
        "range": Array [35, 37],
        "raw": "cl",
        "rawValue": "CL",
        "type": "Register",
        "value": 2,
      },
    ],
    "range": Array [25, 37],
  },
  Object {
    "instruction": Object {
      "mnemonic": "MOV",
      "opcode": 211,
      "range": Array [38, 41],
    },
    "label": null,
    "machineCodes": Array [211, 3, 0],
    "operands": Array [
      Object {
        "range": Array [42, 44],
        "raw": "dl",
        "rawValue": "DL",
        "type": "Register",
        "value": 3,
      },
      Object {
        "range": Array [46, 50],
        "raw": "[al]",
        "rawValue": "AL",
        "type": "RegisterAddress",
        "value": 0,
      },
    ],
    "range": Array [38, 50],
  },
  Object {
    "instruction": Object {
      "mnemonic": "MOV",
      "opcode": 212,
      "range": Array [51, 54],
    },
    "label": null,
    "machineCodes": Array [212, 1, 2],
    "operands": Array [
      Object {
        "range": Array [55, 59],
        "raw": "[bl]",
        "rawValue": "BL",
        "type": "RegisterAddress",
        "value": 1,
      },
      Object {
        "range": Array [61, 63],
        "raw": "cl",
        "rawValue": "CL",
        "type": "Register",
        "value": 2,
      },
    ],
    "range": Array [51, 63],
  },
  Object {
    "instruction": Object {
      "mnemonic": "END",
      "opcode": 0,
      "range": Array [64, 67],
    },
    "label": null,
    "machineCodes": Array [0],
    "operands": Array [],
    "range": Array [64, 67],
  },
]
`;
